### PR TITLE
change(chain): Remove `Copy` trait impl from `Network`

### PR DESCRIPTION
**Pull request by idky137**
## Motivation
This PR continues work towards implementing Regtest mode in Zebrad.

Removing Derive(copy) from Network is required for adding a NetworkParameters struct to Testnet. This breaks a lot of functions that take Network as an input. This PR changes these functions to take &Network where possible and clones Network where either not possible or requires significant refactoring to avoid.

Prerequisite of #7968 and part 3 of #7845.

#### Check before marking the PR as ready for review:
  - [x] Will the PR name make sense to users?
  - [x] Does the PR have a priority label?
  - [x] Have you added or updated tests?
  - [x] Is the documentation up to date?

##### For significant changes:
  - [x] Is there a summary in the CHANGELOG?
  - [x] Can these changes be split into multiple PRs?

_If a checkbox isn't relevant to the PR, mark it as done._

## Solution
Removes Derive(Copy) from Network and any structs that contain Network.
Changes functions that take Network to take &Network where possible and clones network where not.

### Reviewer Checklist

Check before approving the PR:
  - [x] Does the PR scope match the ticket?
  - [x] Are there enough tests to make sure it works? Do the tests cover the PR motivation?
  - [x] Are all the PR blockers dealt with?
        PR blockers can be dealt with in new tickets or PRs.

_And check the PR Author checklist is complete._

### CODE REVIEW

**Comment by idky137** [comment]
Not yet complete. All tests pass in zebra-chain but work still ongoing to fix other crates.

File: zebra-chain/src/block/arbitrary.rs [hunk 1]
```diff
@@ -348,11 +348,12 @@ impl Arbitrary for Block {
     type Parameters = LedgerState;
 
     fn arbitrary_with(ledger_state: Self::Parameters) -> Self::Strategy {
+        let ledger_state_clone = ledger_state.clone();
```
Comment (arya2) [inline zebra-chain/src/block/arbitrary.rs]: Nitpick:

```suggestion
        let transactions_strategy = {
            let ledger_state = ledger_state.clone();
            // Generate a random number transactions. A coinbase tx is always generated, so if
            // `transaction_count` is zero, the block will contain only the coinbase tx.
            (0..MAX_ARBITRARY_ITEMS).prop_flat_map(move |transaction_count| {
                Transaction::vec_strategy(ledger_state, transaction_count)
                Transaction::vec_strategy(ledger_state.clone(), transaction_count)
            })
        };
```

File: zebra-chain/src/block.rs [hunk 2]
```diff
@@ -121,10 +121,10 @@ impl Block {
     #[allow(clippy::unwrap_in_result)]
     pub fn check_transaction_network_upgrade_consistency(
         &self,
-        network: Network,
+        network: &Network,
     ) -> Result<(), error::BlockError> {
         let block_nu =
-            NetworkUpgrade::current(network, self.coinbase_height().expect("a valid height"));
+            NetworkUpgrade::current(&network, self.coinbase_height().expect("a valid height"));
```
Comment (arya2) [inline zebra-chain/src/block.rs]: This looks like a double borrow.

File: zebra-chain/src/block/commitment.rs [hunk 3]
```diff
@@ -103,19 +103,19 @@ impl Commitment {
...
 
-        match NetworkUpgrade::current(network, height) {
+        match NetworkUpgrade::current(&network, height) {
             Genesis | BeforeOverwinter | Overwinter => Ok(PreSaplingReserved(bytes)),
             Sapling | Blossom => match sapling::tree::Root::try_from(bytes) {
                 Ok(root) => Ok(FinalSaplingRoot(root)),
                 _ => Err(InvalidSapingRootBytes),
             },
-            Heartwood if Some(height) == Heartwood.activation_height(network) => {
+            Heartwood if Some(height) == Heartwood.activation_height(&network) => {
```
Comment (arya2) [inline zebra-chain/src/block/commitment.rs]: These look like double borrows too.

File: zebra-chain/src/history_tree.rs [hunk 4]
```diff
@@ -78,12 +78,12 @@ impl NonEmptyHistoryTree {
...
     /// [`NonEmptyHistoryTree::peaks`] and [`NonEmptyHistoryTree::current_height`] of a HistoryTree.
     pub fn from_cache(
-        network: Network,
+        network: &Network,
         size: u32,
         peaks: BTreeMap<u32, Entry>,
         current_height: Height,
     ) -> Result<Self, HistoryTreeError> {
-        let network_upgrade = NetworkUpgrade::current(network, current_height);
+        let network_upgrade = NetworkUpgrade::current(&network, current_height);
```
Comment (arya2) [inline zebra-chain/src/history_tree.rs]: Looks like a double borrow.

File: zebra-chain/src/history_tree.rs [hunk 5]
```diff
@@ -130,15 +130,15 @@ impl NonEmptyHistoryTree {
...
+        network: &Network,
         block: Arc<Block>,
         sapling_root: &sapling::tree::Root,
         orchard_root: &orchard::tree::Root,
     ) -> Result<Self, HistoryTreeError> {
         let height = block
             .coinbase_height()
             .expect("block must have coinbase height during contextual verification");
-        let network_upgrade = NetworkUpgrade::current(network, height);
+        let network_upgrade = NetworkUpgrade::current(&network, height);
```
Comment (arya2) [inline zebra-chain/src/history_tree.rs]: Looks like a double borrow.

File: zebra-chain/src/history_tree.rs [hunk 6]
```diff
@@ -239,8 +239,8 @@ impl NonEmptyHistoryTree {
 
     /// Extend the history tree with the given blocks.
     pub fn try_extend<
-        'a,
-        T: IntoIterator<Item = (Arc<Block>, &'a sapling::tree::Root, &'a orchard::tree::Root)>,
+        'b,
+        T: IntoIterator<Item = (Arc<Block>, &'b sapling::tree::Root, &'b orchard::tree::Root)>,
```
Comment (arya2) [inline zebra-chain/src/history_tree.rs]: Nitpick: This change looks unnecessary, and I'd initially thought there might be another lifetime `'a` in the impl block, so it could be clearer to leave it as `'a`.

```suggestion
        'a,
        T: IntoIterator<Item = (Arc<Block>, &'a sapling::tree::Root, &'a orchard::tree::Root)>,
```

File: zebra-chain/src/history_tree.rs [hunk 7]
```diff
@@ -392,7 +392,7 @@ impl Clone for NonEmptyHistoryTree {
             ),
             InnerHistoryTree::OrchardOnward(_) => InnerHistoryTree::OrchardOnward(
                 Tree::<OrchardOnward>::new_from_cache(
-                    self.network,
+                    &self.network.clone(),
```
Comment (arya2) [inline zebra-chain/src/history_tree.rs]: Nitpick: It may not be necessary to clone here (though cloning an `Arc` is very cheap, so it's not very important either way.)

File: zebra-chain/src/history_tree.rs [hunk 8]
```diff
@@ -414,22 +414,28 @@ impl Clone for NonEmptyHistoryTree {
 
 /// A History Tree that keeps track of its own creation in the Heartwood
 /// activation block, being empty beforehand.
-#[derive(Debug, Default, Clone)]
+#[derive(Debug, Default)]
```
Comment (arya2) [inline zebra-chain/src/history_tree.rs]: Nitpick: Could we restore the derived `Clone` impl?

File: zebra-chain/src/history_tree.rs [hunk 9]
```diff
@@ -414,22 +414,28 @@ impl Clone for NonEmptyHistoryTree {
...
     pub fn from_block(
-        network: Network,
+        network: &Network,
         block: Arc<Block>,
         sapling_root: &sapling::tree::Root,
         orchard_root: &orchard::tree::Root,
     ) -> Result<Self, HistoryTreeError> {
         let heartwood_height = NetworkUpgrade::Heartwood
-            .activation_height(network)
+            .activation_height(&network)
```
Comment (arya2) [inline zebra-chain/src/history_tree.rs]: Looks like a double borrow, same with the borrow below on line 447.

File: zebra-chain/src/history_tree.rs [hunk 10]
```diff
@@ -450,13 +457,13 @@ impl HistoryTree {
...
+        network: &Network,
         block: Arc<Block>,
-        sapling_root: sapling::tree::Root,
-        orchard_root: orchard::tree::Root,
+        sapling_root: &sapling::tree::Root,
+        orchard_root: &orchard::tree::Root,
     ) -> Result<(), HistoryTreeError> {
         let heartwood_height = NetworkUpgrade::Heartwood
-            .activation_height(network)
+            .activation_height(&network)
```
Comment (arya2) [inline zebra-chain/src/history_tree.rs]: Looks like a double borrow.

File: zebra-chain/src/history_tree.rs [hunk 11]
```diff
@@ -472,7 +479,7 @@ impl HistoryTree {
             std::cmp::Ordering::Equal => {
                 let tree = Some(NonEmptyHistoryTree::from_block(
                     network,
-                    block.clone(),
+                    block,
```
Comment (arya2) [inline zebra-chain/src/history_tree.rs:475]: Nice catch!

File: zebra-chain/src/block/commitment.rs (hunk 3 above)
Comment (idky137) [reply zebra-chain/src/block/commitment.rs]: change made

File: zebra-chain/src/block/arbitrary.rs (hunk 1 above)
Comment (idky137) [reply zebra-chain/src/block/arbitrary.rs]: change made

File: zebra-chain/src/block.rs (hunk 2 above)
Comment (idky137) [reply zebra-chain/src/block.rs]: change made

File: zebra-chain/src/history_tree.rs (hunk 4 above)
Comment (idky137) [reply zebra-chain/src/history_tree.rs]: change made

File: zebra-chain/src/history_tree.rs (hunk 5 above)
Comment (idky137) [reply zebra-chain/src/history_tree.rs]: change made

File: zebra-chain/src/history_tree.rs (hunk 6 above)
Comment (idky137) [reply zebra-chain/src/history_tree.rs]: change made (sorry cant quite remember why i did this)

File: zebra-chain/src/history_tree.rs (hunk 7 above)
Comment (idky137) [reply zebra-chain/src/history_tree.rs]: removed

File: zebra-chain/src/history_tree.rs (hunk 8 above)
Comment (idky137) [reply zebra-chain/src/history_tree.rs]: restored

File: zebra-chain/src/history_tree.rs (hunk 9 above)
Comment (idky137) [reply zebra-chain/src/history_tree.rs]: changes made

File: zebra-chain/src/history_tree.rs (hunk 10 above)
Comment (idky137) [reply zebra-chain/src/history_tree.rs]: change made

**Comment by idky137** [comment]
sorry for all the double borrows. i knew i had a couple i've been meaning to go back to but did realise how many

File: zebra-chain/src/history_tree.rs [hunk 12]
```diff
@@ -382,7 +382,7 @@ impl Clone for NonEmptyHistoryTree {
         let tree = match self.inner {
             InnerHistoryTree::PreOrchard(_) => InnerHistoryTree::PreOrchard(
                 Tree::<PreOrchard>::new_from_cache(
-                    self.network,
+                    &self.network.clone(),
```
Comment (arya2) [inline zebra-chain/src/history_tree.rs]: Nitpick: It may not be necessary to clone here.

File: zebra-chain/src/primitives/address.rs [hunk 13]
```diff
@@ -185,7 +185,7 @@ impl Address {
             Self::Transparent(address) => Some(address.to_string()),
             Self::Sapling { address, network } => {
                 let data = address.to_bytes();
-                let address = ZcashAddress::from_sapling((*network).into(), data);
+                let address = ZcashAddress::from_sapling((network).into(), data);
```
Comment (arya2) [inline zebra-chain/src/primitives/address.rs]:
```suggestion
                let address = ZcashAddress::from_sapling(network.into(), data);
```

File: zebra-chain/src/primitives/zcash_history.rs [hunk 14]
```diff
@@ -106,12 +106,12 @@ impl<V: Version> Tree<V> {
...
     #[allow(clippy::unwrap_in_result)]
     pub fn new_from_cache(
-        network: Network,
+        network: &Network,
         network_upgrade: NetworkUpgrade,
         length: u32,
         peaks: &BTreeMap<u32, Entry>,
         extra: &BTreeMap<u32, Entry>,
-    ) -> Result<Self, io::Error> {
+    ) -> Result<Tree<V>, io::Error> {
```
Comment (arya2) [inline zebra-chain/src/primitives/zcash_history.rs]: Nitpick: Unnecessary change / inconsistent.

```suggestion
    ) -> Result<Self, io::Error> {
```

File: zebra-chain/src/primitives/zcash_history.rs [hunk 15]
```diff
@@ -140,15 +140,15 @@ impl<V: Version> Tree<V> {
...
+        network: &Network,
         block: Arc<Block>,
         sapling_root: &sapling::tree::Root,
         orchard_root: &orchard::tree::Root,
     ) -> Result<(Self, Entry), io::Error> {
         let height = block
             .coinbase_height()
             .expect("block must have coinbase height during contextual verification");
-        let network_upgrade = NetworkUpgrade::current(network, height);
+        let network_upgrade = NetworkUpgrade::current(&network, height);
```
Comment (arya2) [inline zebra-chain/src/primitives/zcash_history.rs]: Looks like a double borrow.

```suggestion
        let network_upgrade = NetworkUpgrade::current(network, height);
```

File: zebra-chain/src/primitives/zcash_history.rs [hunk 16]
```diff
@@ -233,14 +233,14 @@ impl Version for zcash_history::V1 {
...
-        network: Network,
+        network: &Network,
         sapling_root: &sapling::tree::Root,
         _orchard_root: &orchard::tree::Root,
     ) -> Self::NodeData {
         let height = block
             .coinbase_height()
             .expect("block must have coinbase height during contextual verification");
-        let network_upgrade = NetworkUpgrade::current(network, height);
+        let network_upgrade = NetworkUpgrade::current(&network, height);
```
Comment (arya2) [inline zebra-chain/src/primitives/zcash_history.rs]: Looks like a double borrow.

```suggestion
        let network_upgrade = NetworkUpgrade::current(network, height);
```

File: zebra-chain/src/primitives/zcash_history/tests/vectors.rs [hunk 17]
```diff
@@ -11,17 +11,17 @@ use eyre::Result;
...
+    tree_for_network_upgrade(&Network::Testnet, NetworkUpgrade::Canopy)?;
     Ok(())
 }
 
-fn tree_for_network_upgrade(network: Network, network_upgrade: NetworkUpgrade) -> Result<()> {
+fn tree_for_network_upgrade(network: &Network, network_upgrade: NetworkUpgrade) -> Result<()> {
     let (blocks, sapling_roots) = network.block_sapling_roots_map();
 
-    let height = network_upgrade.activation_height(network).unwrap().0;
+    let height = network_upgrade.activation_height(&network).unwrap().0;
```
Comment (arya2) [inline zebra-chain/src/primitives/zcash_history/tests/vectors.rs]: Looks like a double borrow.
```suggestion
    let height = network_upgrade.activation_height(network).unwrap().0;
```

File: zebra-chain/src/primitives/zcash_history/tests/vectors.rs [hunk 18]
```diff
@@ -45,7 +45,7 @@ fn tree_for_network_upgrade(network: Network, network_upgrade: NetworkUpgrade) -
     let sapling_root0 =
         sapling::tree::Root::try_from(**sapling_roots.get(&height).expect("test vector exists"))?;
     let (mut tree, _) =
-        Tree::<V1>::new_from_block(network, block0, &sapling_root0, &Default::default())?;
+        Tree::<V1>::new_from_block(&network, block0, &sapling_root0, &Default::default())?;
```
Comment (arya2) [inline zebra-chain/src/primitives/zcash_history/tests/vectors.rs]: Looks like a double borrow.

```suggestion
        Tree::<V1>::new_from_block(network, block0, &sapling_root0, &Default::default())?;
```

File: zebra-chain/src/primitives/zcash_history/tests/vectors.rs [hunk 19]
```diff
@@ -60,7 +60,7 @@ fn tree_for_network_upgrade(network: Network, network_upgrade: NetworkUpgrade) -
     );
 
     // Check its commitment
-    let commitment1 = block1.commitment(network)?;
+    let commitment1 = block1.commitment(&network)?;
```
Comment (arya2) [inline zebra-chain/src/primitives/zcash_history/tests/vectors.rs]:
```suggestion
    let commitment1 = block1.commitment(network)?;
```

File: zebra-chain/src/primitives/zcash_history/tests/vectors.rs [hunk 20]
```diff
@@ -33,7 +33,7 @@ fn tree_for_network_upgrade(network: Network, network_upgrade: NetworkUpgrade) -
     );
 
     // Check its commitment
-    let commitment0 = block0.commitment(network)?;
+    let commitment0 = block0.commitment(&network)?;
```
Comment (arya2) [inline zebra-chain/src/primitives/zcash_history/tests/vectors.rs]:
```suggestion
    let commitment0 = block0.commitment(network)?;
```

File: zebra-chain/src/primitives/zcash_note_encryption.rs [hunk 21]
```diff
@@ -15,8 +15,8 @@ use crate::{
 ///
 /// If passed a network/height without matching consensus branch ID (pre-Overwinter),
 /// since `librustzcash` won't be able to parse it.
-pub fn decrypts_successfully(transaction: &Transaction, network: Network, height: Height) -> bool {
-    let network_upgrade = NetworkUpgrade::current(network, height);
+pub fn decrypts_successfully(transaction: &Transaction, network: &Network, height: Height) -> bool {
+    let network_upgrade = NetworkUpgrade::current(&network, height);
```
Comment (arya2) [inline zebra-chain/src/primitives/zcash_note_encryption.rs]:
```suggestion
    let network_upgrade = NetworkUpgrade::current(network, height);
```

File: zebra-chain/src/transaction/arbitrary.rs [hunk 22]
```diff
@@ -859,12 +859,12 @@ impl Arbitrary for VerifiedUnminedTx {
...
 pub fn transaction_to_fake_v5(
     trans: &Transaction,
-    network: Network,
+    network: &Network,
     height: block::Height,
 ) -> Transaction {
     use Transaction::*;
 
-    let block_nu = NetworkUpgrade::current(network, height);
+    let block_nu = NetworkUpgrade::current(&network, height);
```
Comment (arya2) [inline zebra-chain/src/transaction/arbitrary.rs]:
```suggestion
    let block_nu = NetworkUpgrade::current(network, height);
```

File: zebra-chain/src/transparent.rs [hunk 23]
```diff
@@ -426,7 +426,7 @@ impl Output {
     /// Return the destination address from a transparent output.
     ///
     /// Returns None if the address type is not valid or unrecognized.
-    pub fn address(&self, network: Network) -> Option<Address> {
-        zcash_primitives::transparent_output_address(self, network)
+    pub fn address(&self, network: &Network) -> Option<Address> {
+        zcash_primitives::transparent_output_address(self, &network)
```
Comment (arya2) [inline zebra-chain/src/transparent.rs]:
```suggestion
        zcash_primitives::transparent_output_address(self, network)
```

File: zebra-consensus/src/block/check.rs [hunk 24]
```diff
@@ -154,7 +154,7 @@ pub fn subsidy_is_valid(block: &Block, network: Network) -> Result<(), BlockErro
     };
 
     let canopy_activation_height = NetworkUpgrade::Canopy
-        .activation_height(network)
+        .activation_height(&network)
```
Comment (arya2) [inline zebra-consensus/src/block/check.rs]:
```suggestion
        .activation_height(network)
```

File: zebra-consensus/src/block/check.rs [hunk 25]
```diff
@@ -289,13 +289,13 @@ pub fn time_is_valid_at(
...
 pub fn merkle_root_validity(
-    network: Network,
+    network: &Network,
     block: &Block,
     transaction_hashes: &[transaction::Hash],
 ) -> Result<(), BlockError> {
     // TODO: deduplicate zebra-chain and zebra-consensus errors (#2908)
     block
-        .check_transaction_network_upgrade_consistency(network)
+        .check_transaction_network_upgrade_consistency(&network)
```
Comment (arya2) [inline zebra-consensus/src/block/check.rs]:
```suggestion
        .check_transaction_network_upgrade_consistency(network)
```

File: zebra-consensus/src/block/subsidy/funding_streams.rs [hunk 26]
```diff
@@ -23,9 +23,9 @@ mod tests;
 /// [7.8]: https://zips.z.cash/protocol/protocol.pdf#subsidies
 pub fn funding_stream_values(
     height: Height,
-    network: Network,
+    network: &Network,
 ) -> Result<HashMap<FundingStreamReceiver, Amount<NonNegative>>, Error> {
-    let canopy_height = Canopy.activation_height(network).unwrap();
+    let canopy_height = Canopy.activation_height(&network).unwrap();
```
Comment (arya2) [inline zebra-consensus/src/block/subsidy/funding_streams.rs]:
```suggestion
    let canopy_height = Canopy.activation_height(network).unwrap();
```

File: zebra-consensus/src/block/subsidy/funding_streams/tests.rs [hunk 27]
```diff
@@ -11,13 +11,13 @@ fn test_funding_stream_values() -> Result<(), Report> {
     let network = Network::Mainnet;
```
Comment (arya2) [inline zebra-consensus/src/block/subsidy/funding_streams/tests.rs]: Optional/Nitpick: We could also update the value of `network` to minimize changes.

```suggestion
    let network = &Network::Mainnet;
```

File: zebra-consensus/src/block/subsidy/general.rs [hunk 28]
```diff
@@ -20,9 +20,9 @@ use crate::{funding_stream_values, parameters::subsidy::*};
 /// [7.8]: https://zips.z.cash/protocol/protocol.pdf#subsidies
 ///
 /// Returns `None` if the divisor would overflow a `u64`.
-pub fn halving_divisor(height: Height, network: Network) -> Option<u64> {
+pub fn halving_divisor(height: Height, network: &Network) -> Option<u64> {
     let blossom_height = Blossom
-        .activation_height(network)
+        .activation_height(&network)
```
Comment (arya2) [inline zebra-consensus/src/block/subsidy/general.rs]:
```suggestion
        .activation_height(network)
```

File: zebra-consensus/src/block/subsidy/general.rs [hunk 29]
```diff
@@ -64,9 +64,9 @@ pub fn halving_divisor(height: Height, network: Network) -> Option<u64> {
 /// `BlockSubsidy(height)` as described in [protocol specification §7.8][7.8]
 ///
 /// [7.8]: https://zips.z.cash/protocol/protocol.pdf#subsidies
-pub fn block_subsidy(height: Height, network: Network) -> Result<Amount<NonNegative>, Error> {
+pub fn block_subsidy(height: Height, network: &Network) -> Result<Amount<NonNegative>, Error> {
     let blossom_height = Blossom
-        .activation_height(network)
+        .activation_height(&network)
```
Comment (arya2) [inline zebra-consensus/src/block/subsidy/general.rs]:
```suggestion
        .activation_height(network)
```

File: zebra-consensus/src/block/subsidy/general.rs [hunk 30]
```diff
@@ -123,14 +123,14 @@ mod test {
...
+        halving_for_network(&Network::Mainnet)?;
+        halving_for_network(&Network::Testnet)?;
 
         Ok(())
     }
 
-    fn halving_for_network(network: Network) -> Result<(), Report> {
-        let blossom_height = Blossom.activation_height(network).unwrap();
+    fn halving_for_network(network: &Network) -> Result<(), Report> {
+        let blossom_height = Blossom.activation_height(&network).unwrap();
```
Comment (arya2) [inline zebra-consensus/src/block/subsidy/general.rs]:
```suggestion
        let blossom_height = Blossom.activation_height(network).unwrap();
```

File: zebra-consensus/src/block/subsidy/general.rs [hunk 31]
```diff
@@ -249,14 +249,14 @@ mod test {
...
+        block_subsidy_for_network(&Network::Mainnet)?;
+        block_subsidy_for_network(&Network::Testnet)?;
 
         Ok(())
     }
 
-    fn block_subsidy_for_network(network: Network) -> Result<(), Report> {
-        let blossom_height = Blossom.activation_height(network).unwrap();
+    fn block_subsidy_for_network(network: &Network) -> Result<(), Report> {
+        let blossom_height = Blossom.activation_height(&network).unwrap();
```
Comment (arya2) [inline zebra-consensus/src/block/subsidy/general.rs]:
```suggestion
        let blossom_height = Blossom.activation_height(network).unwrap();
```

File: zebra-consensus/src/router.rs [hunk 32]
```diff
@@ -242,6 +242,7 @@ where
 
     let checkpoint_state_service = state_service.clone();
     let checkpoint_sync = config.checkpoint_sync;
+    let network_clone = network.clone();
```
Comment (arya2) [inline zebra-consensus/src/router.rs]: Nitpick: Could we move this to a block?

**Review by arya2 — COMMENTED** [review]
This is looking good, thank you for these changes!

There are some double borrows left, and it may be possible to reduce the changes a little by reassigning `network = &network` instead of updating each reference where there are multiple references.

The changes are more substantial than expected, making it a little difficult to review (and I'm sure it's made it a little difficult to write as well). Since the other issues for adding network parameters are blocked on this, we may want to consider hard-coding the values for `Regtest` and using static references to preserve the `Copy` trait impl on `Network` for now rather than leaving them configurable - though we'd still need to remove the `Copy` trait impl on `Network` if we want to support custom/named public `Testnet`s later.

Note: When there are a lot of comments, we often find it helpful to include a link to the commit where a comment is addressed (excluding cases where a Github suggestion has been applied).

**Comment by idky137** [comment]
conversion now complete. all tests pass (workspace wide). I'll go through the review now and make the changes requested. Ye i have been worried that its been getting big but now its complete maybe it will be possible?

File: zebra-chain/src/history_tree.rs (hunk 12 above)
Comment (idky137) [reply zebra-chain/src/history_tree.rs]: change applied

File: zebra-chain/src/primitives/address.rs (hunk 13 above)
Comment (idky137) [reply zebra-chain/src/primitives/address.rs]: change applied

File: zebra-chain/src/primitives/zcash_history.rs (hunk 14 above)
Comment (idky137) [reply zebra-chain/src/primitives/zcash_history.rs]: change applied

File: zebra-chain/src/primitives/zcash_history.rs (hunk 15 above)
Comment (idky137) [reply zebra-chain/src/primitives/zcash_history.rs]: change applied

File: zebra-chain/src/primitives/zcash_history.rs (hunk 16 above)
Comment (idky137) [reply zebra-chain/src/primitives/zcash_history.rs]: change applied

File: zebra-consensus/src/block/subsidy/general.rs (hunk 28 above)
Comment (idky137) [reply zebra-consensus/src/block/subsidy/general.rs]: change applied

File: zebra-chain/src/primitives/zcash_history/tests/vectors.rs (hunk 17 above)
Comment (idky137) [reply zebra-chain/src/primitives/zcash_history/tests/vectors.rs]: change applied

File: zebra-chain/src/primitives/zcash_history/tests/vectors.rs (hunk 18 above)
Comment (idky137) [reply zebra-chain/src/primitives/zcash_history/tests/vectors.rs]: change applied

File: zebra-chain/src/primitives/zcash_history/tests/vectors.rs (hunk 19 above)
Comment (idky137) [reply zebra-chain/src/primitives/zcash_history/tests/vectors.rs]: change applied

File: zebra-chain/src/primitives/zcash_history/tests/vectors.rs (hunk 20 above)
Comment (idky137) [reply zebra-chain/src/primitives/zcash_history/tests/vectors.rs]: change applied

File: zebra-chain/src/primitives/zcash_note_encryption.rs (hunk 21 above)
Comment (idky137) [reply zebra-chain/src/primitives/zcash_note_encryption.rs]: change applied

File: zebra-chain/src/transaction/arbitrary.rs (hunk 22 above)
Comment (idky137) [reply zebra-chain/src/transaction/arbitrary.rs]: change applied

File: zebra-chain/src/transparent.rs (hunk 23 above)
Comment (idky137) [reply zebra-chain/src/transparent.rs]: change applied

File: zebra-consensus/src/block/check.rs (hunk 24 above)
Comment (idky137) [reply zebra-consensus/src/block/check.rs]: change applied

File: zebra-consensus/src/block/check.rs (hunk 25 above)
Comment (idky137) [reply zebra-consensus/src/block/check.rs]: change applied

File: zebra-consensus/src/block/subsidy/funding_streams.rs (hunk 26 above)
Comment (idky137) [reply zebra-consensus/src/block/subsidy/funding_streams.rs]: change applied

File: zebra-consensus/src/block/subsidy/funding_streams/tests.rs (hunk 27 above)
Comment (idky137) [reply zebra-consensus/src/block/subsidy/funding_streams/tests.rs]: change applied

File: zebra-consensus/src/block/subsidy/general.rs (hunk 28 above)
Comment (idky137) [reply zebra-consensus/src/block/subsidy/general.rs]: change applied

File: zebra-consensus/src/block/subsidy/general.rs (hunk 29 above)
Comment (idky137) [reply zebra-consensus/src/block/subsidy/general.rs]: change applied

File: zebra-consensus/src/block/subsidy/general.rs (hunk 30 above)
Comment (idky137) [reply zebra-consensus/src/block/subsidy/general.rs]: change applied

File: zebra-consensus/src/block/subsidy/general.rs (hunk 31 above)
Comment (idky137) [reply zebra-consensus/src/block/subsidy/general.rs]: change applied

File: zebra-consensus/src/router.rs (hunk 32 above)
Comment (idky137) [reply zebra-consensus/src/router.rs]: I wasn't sure whether you want this in its own function or just in a block in this function. Changed to: 
```
let (checkpoint_state_service, checkpoint_sync, network_clone) = {
        let checkpoint_state_service = state_service.clone();
        let checkpoint_sync = config.checkpoint_sync;
        let network_clone = network.clone();

        (checkpoint_state_service, checkpoint_sync, network_clone)
    };
```

**Comment by idky137** [comment]
all clippy warnings fixed and changes from review applied

**Comment by idky137** [comment]
sorry I just saw your comment about linking the commit that addresses a comment. I'll make sure to include this from now on.

**Comment by arya2** [comment]
Minor build failure: https://github.com/ZcashFoundation/zebra/actions/runs/8344331467/job/22837074445?pr=8354#step:12:4138

**Comment by idky137** [comment]
fixed build error (in commit: b09707ca4a1b38200287a1bdecef091ae6b3ac5b)

File: zebra-state/src/service/finalized_state/zebra_db.rs [hunk 33]
```diff
@@ -209,13 +209,13 @@ impl ZebraDb {
...
             new_version,
-            self.network(),
+            &self.network(),
         )
     }
 
     /// Returns the configured network for this database.
     pub fn network(&self) -> Network {
-        self.db.network()
+        self.db.network().clone()
```
Comment (arya2) [inline zebra-state/src/service/finalized_state/zebra_db.rs:218]: No change needed / Nitpick: This looks like it's cloning twice, but that's not important or a blocker, so let's get this merged and I'll update it in the next PR. Just leaving this as a reminder to myself.

File: zebra-state/src/service/non_finalized_state/chain.rs [hunk 34]
```diff
@@ -1270,8 +1270,7 @@ impl Chain {
     ) -> impl Iterator<Item = &TransparentTransfers> {
         addresses
             .iter()
-            .copied()
```
Comment (arya2) [inline zebra-state/src/service/non_finalized_state/chain.rs:1273]: Nice catch!

**Review by arya2 — APPROVED** [review]
Let's merge it!

Thank you for the extensive changes.

**Comment by arya2** [comment]
Once it stabilizes, [`Ref::leak`](https://doc.rust-lang.org/std/cell/struct.Ref.html#method.leak) may be useful for similar cases in the future, but for now, I'm glad we were able to avoid [`OnceCell`](https://doc.rust-lang.org/std/cell/struct.OnceCell.html) or being constrained to hard-coded constants. Thanks again.
